app/main.py

```diff
@@ -1,34 +1,215 @@
-        pass
-
-    def fire(self, location: tuple):
-        # This function should check whether the location
-        # is a key in the `self.field`
-        # If it is, then it should check if this cell is the last alive
-        # in the ship or not.
-        pass
+    def __init__(self, ships: list[tuple]) -> None:
+        self.ships = ships
+        self.field = dict()
+        self.create_ships()
+        if not self._validate_field():
+            print("You should re_create your ships.")
+            self.field = dict()
+
+    def fire(self, location: tuple) -> str:
+        if location in self.field:
+            ship = self.field.get(location)
+            ship.fire(row=location[0], column=location[1])
+            if ship.is_drowned:
+                return "Sunk!"
+            return "Hit!"
+        return "Miss!"
+
+    def create_ships(self) -> None:
+        ships = []
+        for location in self.ships:
+            ships.append(Ship(start=location[0], end=location[1]))
+
+        for ship in ships:
+            for deck in ship.decks:
+                self.field[(deck.row, deck.column)] = ship
+
+    def print_field(self) -> None:
+        whole_field = dict()
+        for row_index in range(10):
+            for column_index in range(10):
+                symbol = "~"
+                if (row_index, column_index) in self.field:
+                    symbol = "□"
+                whole_field[(row_index, column_index)] = symbol
+
+        for location in whole_field:
+            ship = self.field.get(location)
+            if ship:
+                symbol = "*"
+                if ship.is_drowned:
+                    symbol = "x"
+                for deck in ship.decks:
+                    if not deck.is_alive:
+                        whole_field[(deck.row, deck.column)] = symbol
+
+        for index, location in enumerate(whole_field):
+            if index % 10 == 0:
+                print("")
+            print(f"{whole_field.get(location)}      ", end="")
+
+    def _validate_field(self) -> bool:
+        count_ships = dict()
+
+        ships = set(self.field.values())
+        for ship in ships:
+            ship_length = len(ship.decks)
+
+            if count_ships.get(ship_length):
+                count_ships[ship_length] += 1
+                continue
+            count_ships[ship_length] = 1
+
+        checks = (
+            count_ships.get(1) == 4,
+            count_ships.get(2) == 3,
+            count_ships.get(3) == 2,
+            count_ships.get(4) == 1,
+            self.check_neighbours_cells(ships)
+        )
+
+        if all(checks):
+            return True
+        return False
+
+    def check_neighbours_cells(self, ships: set[Ship]) -> bool:
+        for ship in ships:
+            locations = []
+            for deck in ship.decks:
+                locations.append((deck.row, deck.column))
+
+            for location in locations:
+                combinations = [
```

Comment:
You can declare a variable `combinations` at the beginning of a function and not rewrite it with each iteration.